infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/RuntimeHangingDetector.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes;
+
+import static java.util.Collections.emptyMap;
+
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.RuntimeEventsPublisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tracks STARTING/STOPPING runtimes and forcibly stop them if they did not change status before a
+ * timeout is reached.
+ *
+ * @author Sergii Leshchenko
+ */
+@Singleton
+public class RuntimeHangingDetector {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHangingDetector.class);
+
+  private Timer timeouts;
+
+  private final RuntimeEventsPublisher eventPublisher;
+  private final Map<String, WaitStatusChangedTask> workspaceId2Task;
+
+  @Inject
+  public RuntimeHangingDetector(RuntimeEventsPublisher eventPublisher) {
+    this.eventPublisher = eventPublisher;
+    this.workspaceId2Task = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Should be invoked after recovering of {@link WorkspaceStatus#STARTING} runtime to track it and
```

Comment:
Could you also describe what it does, not only when it is needed?

Reply:
Makes sense, I'll update java doc.

Reply:
Done in https://github.com/eclipse/che/pull/11570/commits/158a25488c719f49d7e3c602dcc4fce7a749c009. Let me know if it can be improved.